integration-libs/opf/base/core/services/opf-payment-hosted-fields.service.ts

```diff
@@ -0,0 +1,234 @@
+  OpfOrderFacade,
+  OpfOtpFacade,
+  OpfPaymentError,
+  PaymentErrorType,
+  PaymentMethod,
+  SubmitCompleteInput,
+  SubmitCompleteRequest,
+  SubmitCompleteResponse,
+  SubmitInput,
+  SubmitRequest,
+  SubmitResponse,
+  SubmitStatus,
+} from '@spartacus/opf/base/root';
+import { OpfPaymentConnector } from '../connectors/opf-payment.connector';
+import { OpfPaymentErrorHandlerService } from '../services/opf-payment-error-handler.service';
+import {
+  isAuthorizationError,
+  opfAuthorizationErrorRetry,
+} from '../utils/opf-occ-http-error-handlers';
+import { getBrowserInfo } from '../utils/opf-payment-utils';
+
+@Injectable()
+export class OpfPaymentHostedFieldsService {
+  constructor(
+    protected opfPaymentConnector: OpfPaymentConnector,
+    protected winRef: WindowRef,
+    protected opfOtpFacade: OpfOtpFacade,
+    protected activeCartFacade: ActiveCartFacade,
+    protected userIdService: UserIdService,
+    protected routingService: RoutingService,
+    protected opfOrderFacade: OpfOrderFacade,
+    protected globalMessageService: GlobalMessageService,
+    protected opfPaymentErrorHandlerService: OpfPaymentErrorHandlerService
+  ) {}
+
+  submitPayment(submitInput: SubmitInput): Observable<boolean> {
+    const {
+      paymentMethod,
+      cartId,
+      additionalData,
+      paymentSessionId,
+      returnPath,
+      encryptedToken,
+    } = submitInput;
+
+    const submitRequest: SubmitRequest = {
+      paymentMethod,
+      cartId,
+      additionalData,
+      channel: 'BROWSER',
+      browserInfo: getBrowserInfo(this.winRef.nativeWindow),
+    };
+    if (paymentMethod !== PaymentMethod.CREDIT_CARD) {
+      submitRequest.encryptedToken = '';
+    }
+    if (encryptedToken) {
+      submitRequest.encryptedToken = encryptedToken;
+    }
+
+    return combineLatest([
+      this.userIdService.getUserId(),
+      this.activeCartFacade.getActiveCartId(),
+    ]).pipe(
+      filter(
+        ([userId, activeCartId]: [string, string]) => !!activeCartId && !!userId
+      ),
+      switchMap(([userId, activeCartId]: [string, string]) => {
+        submitRequest.cartId = activeCartId;
+        return this.opfOtpFacade.generateOtpKey(userId, activeCartId);
+      }),
+      take(1),
+      concatMap(({ accessCode: otpKey }) =>
+        this.opfPaymentConnector.submitPayment(
+          submitRequest,
+          otpKey,
+          paymentSessionId
+        )
+      ),
+      concatMap((response: SubmitResponse) =>
+        this.paymentResponseHandler(response, submitInput.callbackArray)
+      ),
+      tap((order: Order) => {
+        if (order) {
+          this.routingService.go({ cxRoute: 'orderConfirmation' });
+        }
+      }),
+      map((order: Order) => (order ? true : false)),
+      catchError((error: OpfPaymentError | undefined) => {
+        this.opfPaymentErrorHandlerService.handlePaymentError(
+          error,
+          returnPath
+        );
+        return throwError(error);
+      }),
```

Comment:
[NITPICK]

No need to catch and then re-throw, becasue we don't map the error to something elese.
More semantically would be to call a  `tap({error})`. See example:

```suggestion
      tap({
        error: (error: OpfPaymentError | undefined) => {
          this.opfPaymentErrorHandlerService.handlePaymentError(
            error,
            returnPath
          );
        },
      }),
```